Source/WebCore/rendering/RenderLayerBacking.cpp

```diff
@@ -2913,6 +2928,43 @@ bool RenderLayerBacking::isDirectlyCompositedImage() const
     return false;
 }
 
+bool RenderLayerBacking::isUnscaledBitmapOnly() const
+{
+    if (!is<RenderImage>(renderer()) && !is<RenderHTMLCanvas>(renderer()))
+        return false;
+
+    if (m_owningLayer.hasVisibleBoxDecorationsOrBackground())
+        return false;
+
+    auto contents = contentsBox();
+    if (contents.location() != LayoutPoint(0, 0))
+        return false;
+
+    if (is<RenderImage>(renderer())) {
+        // FIXME: Test image path.
+        auto& imageRenderer = downcast<RenderImage>(renderer());
+        if (auto* cachedImage = imageRenderer.cachedImage()) {
+            if (!cachedImage->hasImage())
+                return false;
+
+            auto* image = cachedImage->imageForRenderer(&imageRenderer);
+            if (!is<BitmapImage>(image))
+                return false;
+
+            if (downcast<BitmapImage>(*image).orientationForCurrentFrame() != ImageOrientation::None)
+                return false;
+
+            return contents.size() == image->size();
```

Comment:
I feel like there needs to be some snapped size comparison here.

Reply:
Relatedly, does an image on a sub pixel horizontal position in a layer render perfectly with the patch? Maybe test with a ref test?

Reply:
We already check if the image is at 0,0 within the layer, and reject this optimisation if not.

I tried creating a test where the layer itself was offset (using position:relative/left:0.5px). That fails since it looks like we paint the non-layer case without snapping (and antialias the edges), and composite the layer case snapped (with sharp edges). I also tried a transformed equivalent.